SUSE frontend challenge (Patrícia Espada)

patriciaespada opened the pull request:
Hi,

Here is my submission of the frontend challenge. Any questions/doubts just let me know.

Kind regards,
Patrícia Espada

Commit:
Create react app.

Commit:
Added Bootstrap support, and created a preliminary layout.

Commit:
Separe application into different components.

Commit:
Added support to CSS Preprocessor.

Commit:
Move static content.

Commit:
Add a first version of the datatable with the short list of computer systems.

Commit:
Fix the package json, the nom-run-all dependency was missing.

Commit:
Add the expandable information to each table row.

Commit:
Added new component that deals with presenting a row more detailed information.

Commit:
Display additional details for the row selected.

Commit:
Some responsive adjustments.

Commit:
Make the code more clean and readable. Include some documentation.

Commit:
Fix indentation format in all files.

Commit:
Add server application to provide an api to retrieve json files information.

Commit:
Refactor client code into specific folder.

Commit:
Request the server to get the computer systems list.

Commit:
Call the server to receive the details about a specific computer system. Abstract the server calls into a separated file..

Commit:
The get all computer systems now manipulates the json to add a random provision state (running, pending, stopped).

Commit:
Add graph component to show the computer system status: running, pending or stopped.

Commit:
Add README file and fix some javascript bugs.

Commit:
Some adjustments to support prod env.

cavalheiro commented:
Hello Patrícia, thank you for applying!
Can you please add @ncounter and @mateiw as reviewers?

patriciaespada commented:
Hi João,
Just send a collaboration invite to both @ncounter and @mateiw. As soon as they accept I will add them as reviewers.

mateiw commented on client/src/components/ComputerSysDescription.js:
```diff
@@ -0,0 +1,42 @@
+/**
+ * Component responsible to present the basic information of the selected computer system.
+ */
+class ComputerSysDescription extends Component {
```
Nitpick: I think this could be a stateless component.

patriciaespada replied:
Also correct. Going to change this stateless components observations that you made.

mateiw commented on client/src/components/MainHeader.js:
```diff
@@ -0,0 +1,19 @@
+		return (
+			<PageHeader bsClass="page-header">
+				Computer Systems Listing App
+				<br/><small>Simple UI for listing computer systems and their specifications.</small>
```
How would you approach the problem of supporting multiple languages for the UI ?

patriciaespada replied:
First I would store the language support in the server side with all the needed translations for every language. Then, upon request I would return a JSON file with all translations for a specific language, with key values pairs that describe each string I would display in the UI.

On the client side I would create a top level translation component that would map the information received from the server with the needs for each component. So whenever a component is rendering it would request the translation component for a particular translation, with a specific key. The translator component just need to match that key with something received on the JSON file.

Of course we would need to have in the UI a way to select the supported languages.

mateiw commented on client/src/components/ComputerSysMoreDetails.js:
```diff
@@ -0,0 +1,118 @@
+	componentWillReceiveProps(nextProps) {
+		if (nextProps.data !== this.props.data) {
+			this.setState({isLoading: true});
+			this.getComputerSystemDetails(nextProps.data);
```
Since this triggers an sync call to fetch the data (`fetchComputerSystemDetail`) would it be needed to have some spinner or something to show the data is loading ?

patriciaespada replied:
I have also added a spinner for this case. It's down on this component when we do the render.

mateiw commented on client/src/components/ComputerSysMoreDetails.js:
```diff
@@ -0,0 +1,118 @@
+	 * @param string id 
+	 */
+	getComputerSystemDetails(id) {
+		fetchComputerSystemDetails(id)
```
I think another option would be for the data fetching to be done in a top level component and to pass a callback here.

patriciaespada replied:
Yes, I believe I can do that improvement.

mateiw commented on client/src/js/ProvisionStateChartUtil.js:
```diff
@@ -0,0 +1,24 @@
+	var pending = 0;
+	var running = 0;
+	var stopped = 0;
+	computerSystems.forEach(function (computerSystem) {
```
An arrow function could be used here.

patriciaespada replied:
Right again. I'm going to fix this arrow function tips. Thanks.

mateiw commented on client/src/components/ComputerSysDescription.js:
```diff
@@ -0,0 +1,42 @@
+	 * @param int to 
+	 */
+	renderFragment(computerSystem, from, to) {
+		return Object.keys(computerSystem).map(function(key, index) {
```
An arrow function should also do the trick here.

patriciaespada replied:
True, we can make this better.

mateiw commented on client/src/components/NavigationBar.js:
```diff
@@ -0,0 +1,37 @@
+/**
+ * Create the navigation bar.
+ */
+class NavigationBar extends Component {
```
This could also be a stateless component.

patriciaespada replied:
And again true here. Thanks for pointing it out.

mateiw commented on client/src/components/ComputerSysContainer.js:
```diff
@@ -0,0 +1,27 @@
+/**
+ * Component to show the selected computer system information.
+ */
+class ComputerSysContainer extends Component {
```
Nitpick: This could be stateless.

patriciaespada replied:
Yes, I believe you are correct. Since this is just a presentation component it doesn't make sense not to be a stateless one.

mateiw reviewed (changes requested):
First of all thank you for your contribution. I added a few comments, mostly minor things.
One global question, how would you make the UI and API update the server status (stopped/running/pending) in real time ?

patriciaespada commented:
@mateiw thanks for your feedback. It's much appreciated. I will do the improvements you suggest and commit them during the day.

Regarding the question you made above: **how would you make the UI and API update the server status (stopped/running/pending) in real time ?**

I can think of two ways to do this.
A simple solution is to request the server for any changes from time to time, so in my case I would only to set an interval where I want to repeat the api call to the server, again, and again, and again...
A more complex way could be to make the server notify the client that a change occurred on the server status. For this, we would need to configure a websocket on the server side and some kind of schedule process to broadcast the server status information to the client.
On the client side we need also to create a websocket that will connect with the server websocket (we can use react-websocket for that), then we would need to process the received data to correctly display it.

Commit:
Improvements after code review:
- Convert the components that only did the rendering from a state component to a stateless component;
- Convert some functions to arrow functions;
- Do the API requests in a top level component, and the callers pass callback functions to deal with the results.

Commit:
Implement a solution for updating the server status in real-time. This involves creating an websocket on the server side, where the UI then connects and receives all message broadcast.

patriciaespada commented:
@mateiw Already did the improvements you suggest on the commit 9126100. Also took the time to implement a real-time support to update the server status on the UI. Used a websocket mechanism to accomplish that. This is on the commit a5184ab.

ncounter commented on client/src/components/MainContent.js:
```diff
@@ -0,0 +1,114 @@
+				</Row>
+				<Row>
+					<Col mdOffset={1} md={10}>
+						{this.state.isRowSelected ? <ComputerSysContainer data={this.state.row} /> : <strong>Select a computer system from the above table.</strong>}
```
From a usability point of view maybe this could have been placed/behave in a different way: for instance if you have a long table list of 50 elements (but maybe more) the user will end up scrolling back and forth for top to bottom. What do you think?

patriciaespada replied:
True. That is one problem. My first solution was to show the additional information on a hidden line, that will show/hide when a user select one of the rows. But after the implementation I didn't like it because the  datatable was getting to crowded.

To solve this I thought of the solution I implemented. What I believe is that there are two major operation users here (at least): comparation of servers, and see servers details. For the first one, the datatable by it selft works good, we can paginate and sort, and see the major information about each other. For the second one, we don't need many servers to be shown at the datatable and we this the scrolls will not be that long. However, that could still be a problem because it's possible to have a large list and them just what to see a server caracterists and need to scroll down, even if it is just to change the pagination.

In that situation I see two ways to fix that. First, open the detailed information on another page. I personally don't like this solution. The other one, would be to have an go to bottom (or go to the next section) button to lead us right to the next section of the page.

So to summarizing, I would go with the last approach that I specify.

ncounter commented on client/src/components/MainContent.js:
```diff
@@ -0,0 +1,114 @@
+	render() {
+		const {status, computerSystems, isLoading, updateSelectedRow} = this.state;
+
+		if (isLoading) {
```
Maybe the conditional returns pattern and can be wrapped up a bit with some `element variables`.

patriciaespada replied:
True, in that way the return will be more clear and don't need to have multiple returns. Nice tip, thanks.

ncounter commented on client/src/components/MainContent.js:
```diff
@@ -0,0 +1,114 @@
+
+		return (
+			<Grid bsClass="main-content">
+				<Websocket url={WS_URL} onMessage={this.handleData.bind(this)}/>
```
Some `onError` or `onClose` handling would help here to inform the user the data are not up-to-date.

patriciaespada replied:
Agree, that will improve the user experience.

ncounter reviewed (commented):
Hi Patricia, thank you for this PR, the App looks good.
I left you a couple of minor comments.

Generally speaking, what's the main reason you used a Pie chart instead of something different?

ncounter commented on client/src/components/ComputerSysDataTable.js:
```diff
@@ -0,0 +1,108 @@
+					<TableHeaderColumn width='200' dataField='contact_method_id' dataSort>Contact Method ID</TableHeaderColumn>
+					<TableHeaderColumn width='150' dataField='running_kernel' dataSort>Running Kernel</TableHeaderColumn>
+					<TableHeaderColumn width='130' dataField='last_boot' dataFormat={this.tableDateFormat} formatExtraData={"last_boot"} dataSort>Last Boot</TableHeaderColumn>
+					<TableHeaderColumn width='160' dataField='provision_state_id' dataSort>Provision State ID</TableHeaderColumn>
```
It would be nice if some *visual* information like this is **graphically** represented in the table as well.

patriciaespada replied:
I could for example create a graph with an availability average for all the servers across time, using the last_boot information.

patriciaespada commented:
Thanks @ncounter for your feedback. The thinks you point it out are indeed some improvements worth doing. Regarding your question: **Generally speaking, what's the main reason you used a Pie chart instead of something different?**

I really like pie charts, are very clear and easy to analyze. Of course they have their purpose and cannot be used for everything. But normally when you need to arrange data from categories and display percentages they are a good fit. In this case, to show the percentage of servers that are in a specific state, the pie chart seems a good way to view that information.

Commit:
Improvements after second code review:
- Add a scroll mechanism two navigate from different sections of the app;
- Added information about the last update date from the server;
- Instead of having multiple conditional returns, include some element variables.

patriciaespada commented:
@ncounter I did the  improvements you suggest and implement a way to fix the scroll problem. All this is in the commit 71fc559.
